python_modules/dagster/dagster/_core/definitions/asset_graph.py

```diff
@@ -60,6 +60,19 @@ def get_parents(self, asset_key: AssetKey) -> AbstractSet[AssetKey]:
         """Returns all assets that the given asset depends on"""
         return self.asset_dep_graph["upstream"][asset_key]
 
+    def get_roots(self, asset_key: AssetKey) -> AbstractSet[AssetKey]:
+        """Returns all root assets that the given asset depends on"""
+        parents = self.get_parents(asset_key)
+        if not parents:
+            return {asset_key}
+        return set().union(*(self.get_roots(parent_key) for parent_key in parents))
+
+    def upstream_key_iterator(self, asset_key: AssetKey) -> Iterator[AssetKey]:
```

Comment:
See comment here: https://github.com/dagster-io/dagster/pull/10366/files#r1018266953

Reply:
🤦 can't believe I didn't think of BFS for this